kernel/arch/dreamcast/include/arch/irq.h

```diff
@@ -255,60 +278,102 @@ typedef enum irq_exception {
     EXC_UNHANDLED_EXC      = 0x0fe0  /**< `[SOFT  ]` Exception went unhandled */
 } irq_t;
 
+/** \defgroup irq_state     State
+    \brief                  Methods for querying active IRQ information.
 
-/** \defgroup  irq_type_offsets        Exception type offsets
-    \brief                             Offsets within exception types
-    \ingroup                           irqs
-
-    The following are a table of "type offsets" (see the Hitachi PDF). These are
-    the 0x000, 0x100, 0x400, and 0x600 offsets.
+    Provides an API for accessing the state of the current IRQ context such
+    as the active interrupt or whether it has been handled.
 
     @{
 */
-#define EXC_OFFSET_000  0   /**< \brief Offset 0x000 */
-#define EXC_OFFSET_100  1   /**< \brief Offset 0x100 */
-#define EXC_OFFSET_400  2   /**< \brief Offset 0x400 */
-#define EXC_OFFSET_600  3   /**< \brief Offset 0x600 */
-/** @} */
 
-/** \brief   The value of the timer IRQ
-    \ingroup irqs
+/** Returns current interrupt depth.
+
+    Checks how deeply neested the caller is within interrupt handlers.
+
+    \note
+    See \ref irqs for more information on nested interrupts.
+
+    \retval 0                   There is no interrupt active.
+    \retval >0                  Caller is N interrupts deep.
+    
+    \sa irq_inside_int(), irq_active_int()
 */
-#define TIMER_IRQ       EXC_TMU0_TUNI0
+size_t irq_int_depth(void);
 
-/** \defgroup irq_state     State
-    \brief                  Methods for querying active IRQ information.
+/** Returns whether inside of an interrupt context.
 
-    Provides an API for accessing the state of the current IRQ context such
-    as the active interrupt or whether it has been handled.
+    \retval true               If interrupt handling is in progress.
+    \retval false              If normal processing is in progress.
 
-    @{
+    \sa irq_int_depth(), irq_active_int()
 */
+bool irq_inside_int(void);
 
+/** Returns the active IRQ source N levels deep.
 
-/** Returns whether inside of an interrupt context.
+    \note
+    See \ref irqs for more information on nested interrupts.
+
+    \param  level           0-based index for depth of the active IRQ
+                            (0 is the current IRQ, depth-1 is first IRQ).
+
+    \retval >0              Exception code for Nth IRQ context.
+    \retval 0               No IRQ context is active.
+
+    \sa irq_inside_int(), irq_int_depth()
+*/
+irq_t irq_active_int(size_t level);
+
+/** Returns whether the current IRQ has been handled.
+
+    Used to determine whether the active interrupt has been accepted by a
```

Comment:
Isn't it only the active interrupt when level=0?

Reply:
Perhaps I misused the term `active`. In my head, in the case of multiple nested IRQs, I was thinking you'd consider each IRQ in this "stack" as *active*... 

And yeah, right now, it is only useful to ever call with a level of 0; however, the API was written to allow for nested interrupts in the future, so it's future-proof and won't have to break then.

Reply:
Ah, ok, that makes sense. I was thinking 'active' as 'the one the handler is currently processing'. It makes perfect sense though to consider the whole stack as active. I have a feeling the term 'raised' would have a role here, but it makes enough sense as-is.